scripts/prepare_lambda_deployment.py

```diff
@@ -70,6 +77,56 @@ def patch_cloudwatch_resource(resource_type_name, property_name):
     return input_json
 
 
+def patch_resource_names(input_json: JSON) -> JSON:
+    # Some Chalice-generated resources have named ending in `-event`. The
+    # dash prevents generation of a fully qualified resource name (i.e.,
+    # with `config.qualified_resource_name`. Returns dict of old
+    # Terraform names and new Terraform names.
+    rename = {
+        'aws_cloudwatch_event_rule.indexercachehealth-event': 'indexercachehealth',
+        'aws_cloudwatch_event_target.indexercachehealth-event': 'indexercachehealth',
+        'aws_lambda_permission.indexercachehealth-event': 'indexercachehealth',
+        'aws_cloudwatch_event_rule.servicecachehealth-event': 'servicecachehealth',
+        'aws_cloudwatch_event_target.servicecachehealth-event': 'servicecachehealth',
+        'aws_lambda_permission.servicecachehealth-event': 'servicecachehealth',
+    }
+    replace = {k: '.'.join([k.split('.')[0], v]) for k, v in rename.items()}
+    literal = typing.TypeVar('literal', str, int)
+
+    def _patch_name(resource_type: str, resource_name: str) -> str:
+        try:
+            new_name = rename[f'{resource_type}.{resource_name}']
+        except KeyError:
+            return resource_name
+        else:
+            return new_name
+
+    def _replace_refs(value: literal) -> literal:
+        if isinstance(value, str):
+            for oldref, newref in replace.items():
+                value = value.replace(oldref, newref)
+            return value
+        else:
+            return value
+
+    def _patch_refs(block: JSON) -> JSON:
+        return {
+            k: _patch_refs(v) if isinstance(v, dict) else _replace_refs(v)
```

Comment:
I don't it is guaranteed that _replace_refs always gets a str or an int. There should be an assert/require somewhere. That may affect the need for a type var.

Reply:
I left this as-is, as `_replace_refs` will return input unaffected if input is not of type `str`. I think the deficiency here was that `T` was poorly defined, so I have redefined it as

    T = typing.TypeVar('T', str, int, float, bool)  # azul.types.PrimitiveJSON

Reply:
If `block` is `JSON`, then `v` is `AnyJSON`. The conditional only checks for `dict` (incorrectly, BTW, it should be `Mapping`) so that leaves `Sequence` and `PrimitiveJSON` as possible types for the argument to `_replace_refs`. I still think there should be an assert to exclude `Sequence`.

Also, can you think of a way to avoid the brittle coupling between the definition of `T` and PrimitiveJSON?

Reply:
> I still think there should be an assert to exclude `Sequence`.

Ah, I misunderstood previously. I agree with your assessment.

> Also, can you think of a way to avoid the brittle coupling between the definition of `T` and PrimitiveJSON?

My intuition is to unpack `PrimitiveJSON` into the `TypeVar`:

```python
T = typing.TypeVar('T', *PrimitiveJSON.__args__)
```

but PyCharm doesn't like that (`A single constraint is not allowed`). A hacky workaround that PyCharm is okay with is

```python
T = typing.TypeVar('T', *PrimitiveJSON.__args__, typing.NoneType)
```

but that doesn't feel right. Is this the direction I should be looking in?